File: java/challenges/lib/src/main/java/challenges/stacksandqueues/Stack.java
package challenges.stacksandqueues;

import java.util.EmptyStackException;
import java.util.*;

public class Stack {
  public Node top;

  public Stack(){
    this.top = null;
  }

  public void push(Object val){
    Node newNode = new Node(val);
    newNode.next = top;
    top = newNode;
  }

  public Object pop(){
    if (top == null){
      throw new EmptyStackException();
    }
    Node temp = top;
    top = top.next;
    temp.next = null;
    return temp.val;
  }

  public Object peek(){
    if (top == null){
      throw new EmptyStackException();
    }
    return top.val;
  }

  public boolean isEmpty(){
    return top == null;
  }


}

File: javascript/stackAndQueue/stack.js
'use strict';

class Node {
  constructor(val)
  {
    this.val = val;
    this.next = null;
  }
}

class Stack {
  constructor()
  {
    this.top = null;
  }

  push(val) {
    let newNode = new Node(val);

    if (!this.top) {
      this.top = newNode;
    }

    newNode.next = this.top;
    this.top = newNode;
  }

  pop() {
    // try {
    if (!this.top) {
      return null;
    }
    let temp = this.top;
    this.top = top.next;
    temp.next = null;
    return temp.val;
  }
  // catch(e) {
  //   return(e + 'empty stack exception');
  // }
  // }

  peek() {
    return this.top.val;
  }

  isEmpty() {
    return top === null;
  }
}

module.exports = { Stack, Node };
